### ISSUE TITLE
Solve Abacus bias

### ISSUE THREAD

**Issue by maho3** [maho3/ltu-gobig#32]
This is to keep notes on biases appearing when training on LtU-CMASS and predicting on Abacus.

Things that could be trouble:

- Incorrect halo mass definition (1, 5))
- Some large-scale bias (6)
- Data vector is too long (2, 3)
- Positional noising causes bias (7)
- RSDs are wrong (4)

Tests:

1. Train on true Quijote, predict on 1 Gpc/h Abacus
2. PCA pre-compression of data vector
3. Use only equilateral B(k)
4. Train on real-space summaries
5. Train/test using galaxies
6. Train using kmin cut
7. Test constraints vs. noising

**Comment by maho3** [comment]
# 8
Addressed by Sammy here: https://docs.google.com/presentation/d/1tLiTAC5IA_Toqtn8_Ip8RJuHFsnPgt4fdt1jIfxGsSU/edit?usp=sharing

**Comment by maho3** [comment]
## 1
Train on Quijote 1 Gpc/h, Predict on Abacus chopped up into 1 Gpc/h

Using zPk024+zQk0

![Image](https://github.com/user-attachments/assets/489e8eaf-947e-4806-9ff7-05f634b1cd07)

![Image](https://github.com/user-attachments/assets/cb8847c4-d15f-4e93-9648-9d7da9f7d48b)\

Still a bias in sigma8!

* Caveat: Chopping up Abacus into 1 Gpc/h volumes removes the periodic boundaries. However, this should have the impact of reducing power across the boundary, and thus biasing sigma8 low.
* This indicates that the halo definitions are causing the misspecification.

**Comment by maho3** [comment]
After reprocessing abacus halos, using the correct velocities and 'cleaned' halos:

![Image](https://github.com/user-attachments/assets/2ff96add-e52b-40ed-b978-a094b0a1044f)

**Comment by maho3** [comment]
The previous velocity comparison was done on inconsistent cosmologies. This is a closer comparison
Abacus LHID: 136
Quijote LHID: 678
![Image](https://github.com/user-attachments/assets/7e34921b-e3da-4d9f-8ac2-4e42a1653b6a)

And here's the P(k)/B(k) comparison
![Image](https://github.com/user-attachments/assets/47b03bb3-998b-4adf-82ae-aac667969e00)

And the field-level comparison
![Image](https://github.com/user-attachments/assets/8d3eb231-9300-4205-a39a-b82eb0fdb134)

The fields look very similar, but the Pk monopole is far off. I think this decrement at large scale power are the result of us cutting 1 Gpc/h volumes out of the 2 Gpc/h Abacus, and not accounting for non-periodic boundary conditions.

**Comment by maho3** [comment]
Raw prediction on galaxies looks like trash. Training on Quijote and predicting on Abacus 1 Gpc/h using the Zheng07 galaxies. RSD with Pk024+Qk0

![Image](https://github.com/user-attachments/assets/54454a44-a769-4cbf-b545-bad416f6c842)

**Comment by maho3** [comment]
## 1 Gpc/h (Train: Quijote) (Predict: 1 Gpc/h Abacus)
### real-space Pk024

![Image](https://github.com/user-attachments/assets/ae1e2a95-b835-4e78-875c-d4d0655a87ce)

### z-space Pk0

![Image](https://github.com/user-attachments/assets/d6ef3337-5cdf-4363-8c21-496cc8762772)

![Image](https://github.com/user-attachments/assets/0729a13b-6204-4b1c-9184-84ca25c044e8)

### z-space Pk024+Qk0

![Image](https://github.com/user-attachments/assets/20c5741a-e4d2-4d54-b9a7-cd705b0b9a5a)

## 2 Gpc/h (Train: CHARM) (Predict: Abacus)
### z-space Pk024+Qk0
We're still underpredicting sigma8, and now weirdly biasing Omega_m and h predictions
![Image](https://github.com/user-attachments/assets/fde79934-a856-4bdf-8249-7157a9bef1e5)

**Comment by maho3** [comment]
The mean mass is highly sensitive to the cosmology. Here's for a fixed Omega_m how varying sigma8 changes the total number density or mean mass of our top 100k, and vice versa.

![Image](https://github.com/user-attachments/assets/99f74c9b-f478-4bc2-baba-7557c762025c)

![Image](https://github.com/user-attachments/assets/c3368153-2891-488d-89a8-19c64f857eeb)


It seems for some cosmologies (low Omega_m) we don't even have 100k halos. This explains why in some low Omega-m cosmologies the shot noise is significantly different.

![Image](https://github.com/user-attachments/assets/ff1511fc-0e1e-41ec-b604-184e4b97f4bc)

![Image](https://github.com/user-attachments/assets/e6eaec4d-3fb7-40f8-877b-6f4bff83b493)

**Comment by maho3** [comment]
### Different Abacus subvolumes impose some cosmic variance to the Pk0, but not enough to account for differences with Quijote's Pk0

![Image](https://github.com/user-attachments/assets/a305f5fc-acdb-4eaf-9acb-f80d73441944)

![Image](https://github.com/user-attachments/assets/950975a9-c453-4a4e-b40c-437efc00d5f8)

![Image](https://github.com/user-attachments/assets/0fdf4b34-0e5a-4a93-aff8-bd1dd6793a5e)

Notes:
* Correcting for average bulk velocity in each subox didn't help

**Comment by maho3** [comment]
Todo (for myself):
- [ ] Retrain halo inference quijote -> abacus, but only with MAFs
- [ ] Retrain galaxy inference quijote -> abacus, with equally selected abacus halos + shot noise correction
- [ ] Run SimBIG-esque constraining power

**Comment by maho3** [comment]
A similar effect occurs for the raw dark matter density fields. Here is comparing Quijote-448 and Abacus-141

![Image](https://github.com/user-attachments/assets/ff111988-a2a1-4d7c-ba79-c585dc2a7ec8)

**Comment by maho3** [comment]
@adrianbayer suggested that underprediction of sigma8 can be caused by:

- other cosmo params (h, n_s)
- massive neutrinos
- different halo biasing

Here's the distribution of abacus 'base' simulations, flagged by whether they are LCDM or have massive neutrinos:

![Image](https://github.com/user-attachments/assets/5eb62889-58b0-4b29-8956-f80c01fc5ee3)

I am now rerunning my inference tests training on quijote, testing on abacus...

# Halos
## Real-space Pk0
## Real-space Pk024+Qk0
## Redshift-space Pk0
## Redshift-space Pk024+Qk0

# Galaxies
## Real-space Pk0
## Real-space Pk024+Qk0
## Redshift-space Pk0
## Redshift-space Pk024+Qk0

**Comment by maho3** [comment]
Here's a test to see if the velocity discrepancy is a function of halo mass or local density. There is some discrepancy in the high mass/high density regions, but its hard to say whether its physical or the result of slightly different cosmologies.

Abacus-141: [0.345736 0.054281 0.6623   0.9384   0.799575]
Quijote-354: [0.3573  0.05321 0.6891  0.9741  0.8249 ]

![Image](https://github.com/user-attachments/assets/dad5b9b1-9524-44d3-aed6-371d03314a36)
![Image](https://github.com/user-attachments/assets/4ce40602-d43d-4524-9923-6fb71956f212)

**Comment by maho3** [comment]
Here's what happens when we cut out the smallest scales (`kmax=0.1`)

## Halos
### Redshift-space Pk0

![Image](https://github.com/user-attachments/assets/7885b6c3-e862-4071-87b0-b96bf532950c)

### Redshift-space Pk024+Qk0
![Image](https://github.com/user-attachments/assets/a98c489f-66de-497c-a480-d18738243290)

The bias disappears! But almost no constraining power for Pk0

**Comment by maho3** [comment]
Here's what the NLE samples look like, training on quijote and testing on Abacus

![Image](https://github.com/user-attachments/assets/a7ed9920-844a-40ea-ac78-1e0b45872e9a)

![Image](https://github.com/user-attachments/assets/a6bf0712-b3e0-4ba3-840e-8501bf819c34)

![Image](https://github.com/user-attachments/assets/ae008e52-065a-4108-8475-e0fd4c9ecdb4)

**Comment by maho3** [comment]
## NLE Sampling
Here, we train NLE to emulate the likelihood of the redshift-space Pk monopole from the halos, training on Quijote and testing on Abacus.

![Image](https://github.com/user-attachments/assets/115a3e54-c91b-4db7-95f9-12cc0339efcc)
![Image](https://github.com/user-attachments/assets/ca5deb6a-4978-4f45-b10c-830a4da0118b)

There seems to be slightly more suppression of power at small-scales (k>0.3) by the Quijote simulations .

When sampling the posterior p(theta|x), we get the same problems as in the NPE results. Underprediction of sigma8

![Image](https://github.com/user-attachments/assets/01c2be3d-2705-47eb-9305-c0c9172ab7d3)

**Comment by maho3** [comment]
## Zheng07extended galaxy comparison (w/ assem_bias, vel_bias)
Here, we are training on true Quijote galaxies and testing on Abacus galaxies, including a Zheng07 HOD with additional parameters for assembly and velocity bias. This is all within the 1 Gpc/h volume.

### Real-space Pk0

![Image](https://github.com/user-attachments/assets/128d331b-390e-4b2a-941b-240241e3121a)

### Real-space Pk024+Qk0

![Image](https://github.com/user-attachments/assets/1e93caeb-ded4-4a0f-be18-bca748ad71bc)

### Redshift-space Pk0

![Image](https://github.com/user-attachments/assets/3f96fcb9-b62f-44bd-be99-665eed71c8fc)

### Redshift-space Pk024+Qk0

![Image](https://github.com/user-attachments/assets/d56f594e-fa5e-407c-9bbf-68e61af6303a)


Takeaways:
* It seems to work ! The velocity bias (probably)  marginalizes the difference between Quijote/Abacus halo velocities.
* Adding vel/assem bias reduces our constraining power, but we still have non-negligible information on Om and s8. This should improve at larger volumes.
* It's kind of hard to see, but the LCDM + Mnu=0 simulations have tighter posterior variance than those without
* Velocity bias is necessary to describe the difference between quijote/abacus

**Comment by maho3** [comment]
## CHARM confirmation: Zheng07extended galaxy comparison (w/ assem_bias, vel_bias)

Here, we're confirming the results shown above with the Quijote halos, but now with the CHARM halos. Again, training on CHARM, testing on Abacus, in 1 Gpc/h, with Zheng07+assem_bias+vel_bias

### Real-space Pk0

![Image](https://github.com/user-attachments/assets/b8dc7895-8b63-4149-adc7-3dc3a5d420da)

### Real-space Pk024+Qk0

![Image](https://github.com/user-attachments/assets/47c7e699-a345-4190-a106-2943ff8f500b)

### Redshift-space Pk0

![Image](https://github.com/user-attachments/assets/ea2af4d5-5bd6-4f67-ab4f-a6c28650b9b7)

### Redshift-space Pk024+Qk0

![Image](https://github.com/user-attachments/assets/ab97d025-4214-46b8-886c-b1d48ab5c668)

**Comment by maho3** [comment]
## Optuna hyperparameter-tuned, train CHARM, test Abacus
We've adapted our training procedure to now use Optuna. (See https://github.com/maho3/ltu-cmass/pull/115 ) 

We've also run more HODs per abacuslike simulation. There's now 5 HOD's per simulation. All results below are for galaxies, with RSDs.

We gain in constraining power. Interestingly, we are unbiased across most summaries except the full bispectrum. This could perhaps be a result of not sufficient training data, as the Qk0 vector is much longer than Pk024.

![Image](https://github.com/user-attachments/assets/bcbe8848-cf3b-42a0-8468-54ab19d19f6d)

![Image](https://github.com/user-attachments/assets/77578132-7c2b-49a0-aaea-49556a35cd45)

![Image](https://github.com/user-attachments/assets/e215620c-f9b3-4438-ba81-a26d6f4e36b6)

![Image](https://github.com/user-attachments/assets/20a2625d-f6cf-4094-b3c1-38105d89c07c)

**Comment by CompiledAtBirth** [comment]
These are parallel tests with regards to the old noising scheme ("hodzbias" fastpm suite) versus the new noising scheme ("recnoise").

With the old noising sheme (here on comoving box summaries):
- the bias does not appear (or marginally) when we add EqQk0 to the power spectrum vector
- bias appears when we add other triangles than equilateral, and seems to be more severere when just adding IsQk0

Note: notably because the Abacus test set realizations are all from the same initial conditions, the coverage test is not representative.

![Image](https://github.com/user-attachments/assets/12cd5bc9-6f57-4bcc-82b3-c1fca9bb562c)
![Image](https://github.com/user-attachments/assets/bc61f443-4d6a-4db5-a868-e2d54244140b)
![Image](https://github.com/user-attachments/assets/14a68d9b-1ed2-470d-8d2d-587a03db8819)

**Comment by CompiledAtBirth** [comment]
Now the equivalent plots with the new noising scheme, ie  "with a prior on the noise scale that goes as 1/sigma (Jeffrey's prior)", here on sgc_lightcone summaries

- we do not see significant bias when adding isosceles configurations this time
- sigma8 is degenerate with one the the noise parameters (radial ?), hence the expected poorer constraints

![Image](https://github.com/user-attachments/assets/4db48b7c-f9f7-4774-9398-7b387e6294ca)
![Image](https://github.com/user-attachments/assets/fbc6cb22-4175-4a3e-a70a-e88ec5099f48)
![Image](https://github.com/user-attachments/assets/55d71b34-dc98-429c-9c86-5c02f2a6d585)
